module/zfs/zvol.c

```diff
@@ -2239,18 +2236,31 @@ zvol_set_snapdev_sync_cb(dsl_pool_t *dp, dsl_dataset_t *ds, void *arg)
 }
 
 /*
- * Traverse all child snapshot datasets and apply snapdev appropriately.
+ * Traverse all child datasets and apply snapdev appropriately.
+ * We call dsl_prop_set_sync_impl() here to set the value only on the toplevel
```

Comment:
Given the following situation
```
NAME                    PROPERTY  VALUE    SOURCE
testpool                snapdev   hidden   default
testpool/fs             snapdev   hidden   default
testpool/fs/zvol        snapdev   hidden   default
testpool/fs/zvol@snap1  snapdev   -        -
```
> we should set the value for the dataset in question

which translates to `zfs set snapdev=visible testpool`

> and then **we should only be recursing over snapshots if the dataset is a zvol**

which means we should do nothing since `testpool` is not a zvol: we won't create `testpool/fs/zvol@snap1` device node, even if now we have `snapdev=visible` on the zvol:

```
NAME                    PROPERTY  VALUE    SOURCE
testpool                snapdev   visible  local
testpool/fs             snapdev   visible  inherited from testpool
testpool/fs/zvol        snapdev   visible  inherited from testpool
testpool/fs/zvol@snap1  snapdev   -        -
```